test/karma/spec/renewToken.js

```diff
@@ -0,0 +1,258 @@
+/* global jasmine, window, document, URL, Event, Promise */
+require('jasmine-ajax');
+
+var tokens = require('../../util/tokens');
+
+import OktaAuth from '@okta/okta-auth-js';
+import oauthUtil from '../../../lib/oauthUtil';
+import pkce from '../../../lib/pkce';
+import OauthError from '../../../lib/errors/OAuthError';
+
+describe('Renew token', function() {
+
+  const ASSUMED_TIME = 1449699929;
+  const ISSUER = tokens.standardIdTokenParsed.issuer;
+  const CALLBACK_PATH = '/implicit/callback'
+  const REDIRECT_URI = `${ISSUER}${CALLBACK_PATH}`;
+  const CLIENT_ID = tokens.standardIdTokenParsed.clientId;
+  const DEFAULT_CONFIG = {
+    issuer: ISSUER,
+    clientId: CLIENT_ID,
+    redirectUri: REDIRECT_URI,
+  };
+
+  const ACCESS_TOKEN_STR = 'fakeytoken'; // will not be verified in this flow
+  const ACCCESS_TOKEN_PARSED = tokens.standardAccessTokenParsed;
+  const NONCE = tokens.standardIdTokenClaims.nonce;
+  const AUTHORIZATION_CODE = 'FAKEY';
+  const JWKS_URI = 'http://myfake.jwks.local';
+
+  var sdk;
+
+
+  beforeEach(function() {
+    document.body.insertAdjacentHTML('beforeend', '<div id="root"></div>');
+    var date = new Date();
+    date.setTime(ASSUMED_TIME * 1000);
+    jasmine.clock().mockDate(date);
+    jasmine.Ajax.install();
+  });
+
+  afterEach(function() {
+    document.body.removeChild(document.getElementById('root'));
+    jasmine.clock().uninstall();
+    jasmine.Ajax.uninstall();
+  });
+
+  function bootstrap(config) {
+    config = Object.assign({}, DEFAULT_CONFIG, config);
+    sdk = new OktaAuth(config);
+    sdk.tokenManager.clear();
+    return Promise.resolve();
+  }
+
+  function mockWellKnown() {
+    sdk.options.storageUtil.getHttpCache().clearStorage();
+
+    var wellKnown = {
+      'jwks_uri': JWKS_URI,
+      code_challenge_methods_supported: ['S256']
+    };
+    var keys = [
+      tokens.standardKey
+    ];
+    jasmine.Ajax.stubRequest(
+      /.*\.well-known/
+    ).andReturn({
+      status: 200,
+      responseText: JSON.stringify(wellKnown)
+    });
+
+    jasmine.Ajax.stubRequest(
+      JWKS_URI
+    ).andReturn({
+      status: 200,
+      responseText: JSON.stringify({
+        keys: keys
+      })
+    })
+  }
+
+  it('receives/throws error from iframe', function() {
+    // This is the error if requesting scope='offline_access'
```

Comment:
Its possible that someone tries to do an upscope request when renewing a token, but in general this error should be caught during the initial `getWithRedirect` or `getWith*` call. Can we move this to the login flow tests?

Reply:
I was only able to reproduce an error from the backend using the renew method. For getWithRedirect or getWithPopup, it will not accept a scope that does not contain 'openid'. If I set the scope as ['openid', 'offline_access'], the backend gives me the openid stuff and ignores offline_access with no error.

Reply:
>  If I set the scope as ['openid', 'offline_access'], the backend gives me the openid stuff and ignores the offline_access scope with no error

Interesting. This could be a bug. So I'm clear, if we request `['openid', 'offline_access']` scopes during our initial authorize request, we **do not** error out and can still exchange that `authorization_code` for tokens? If so, do you know if there is a `refresh_token` included in that response too?

Reply:
I saw only an access_token and id_token. It seemed to ignore the invalid scope. The error I get when I send only offline_access is something like "all of your scopes are invalid". So I think the backend silently ignores invalid scopes unless all scopes are invalid.

Reply:
I can update the test app so it can more easily be reproduced

Reply:
@danielzeller-okta - Based on our conversation a few weeks back, I'm not sure this is working as expected, as we are still permitting requests through.

From what I understand, this is the **expected** behavior:

- If a browser client requests the `offline_access` scope, Okta will reject the authorize request and return an `invalid_scope` error.

Given that is **not** the case for initial auth, we should investigate why we're seeing a different result.

Reply:
test app has been updated it can accept `scopes` and `responseType` in the URL